### DISCUSSION TITLE
Relationship using primaryjoin for tables in 2 different schema's

### DISCUSSION THREAD

**Discussion by elliotdes** [sqlalchemy/sqlalchemy#11461] [Usage Questions]
Hi,

I am trying to create a relationship between 2 tables that are in different schemas (in the same database). The database is postgres.

```python
class TableB(BaseA):
    __tablename__ = "table_b"
    id: Mapped[int] = mapped_column(primary_key=True)
    composite_id: Mapped[str]

    table_a: Mapped["TableA"] = relationship(
        primaryjoin="TableA.composite_id==TableB.composite_id",
        back_populates="table_b",
    )


class TableA(BaseB):
    __tablename__ = "table_a"

    primary_id: Mapped[int] = mapped_column(primary_key=True)
    id: Mapped[str]
    sequence_id: Mapped[int]

    @hybrid_property
    def composite_id(self):
        return func.concat(self.id, "-", self.sequence_id)

    table_b: Mapped["TableB"] = relationship(
        primaryjoin="TableA.composite_id==TableB.composite_id",
        back_populates="table_a",
    )
```

when I try to use these tables, I get the following error:

> sqlalchemy.exc.InvalidRequestError: When initializing mapper Mapper[TableA(table_a)], expression 'TableA.composite_id==TableB.composite_id' failed to locate a name ("name 'TableB' is not defined"). If this is a class name, consider adding this relationship() to the <class 'app.instruments.models.TableA'> class after both dependent classes have been defined.

Is there a way to resolve this issue?

**Comment by zzzeek** [comment]
the string registry of names is local to a specific declarative base, it's not global.   So if you are using two separate bases (I'd question why that's in use, typically unnecessary) you would need to refer to the other class' attributes explicitly, which you can do within a lambda to resolve cycles:

    primaryjoin=lambda: TableA.composite_id==TableB.composite_id

**Reply by elliotdes** [reply]
Thanks for the reply! Always amazed by the speed and quality of responses in this repo!

I am using a different base for each schema. Would you recommend using one declarative base and defining the schema in each model?

**Reply by zzzeek** [reply]
it's really up to you.  you can have one base but then have sub-bases with different MetaData collections, if you wanted (though that might defeat the string expression for primary join use as well).   or if you share one Base/MetaData, you could have a mixin with the "schema" table argument like this:

```py
class Base(DeclarativeBase):
    pass

class MySchemaBase(Base):
    __abstract__ = True

    __table_args__ = {"schema": "myschema"}

class MyClass(MySchemaBase):
    # ...
```

though then you have to be careful if you put further `__table_args__` on a subclass.

**Reply by elliotdes** [reply]
Thank you! I will go for that solution.

On a separate point, is it possible to define a relationship using `hybrid_property`?

in the above example, this is `composite_id`. When I try to define it in the `foreign_keys` argument, I get the following error:

> sqlalchemy.exc.ArgumentError: Column expression expected for argument 'foreign_keys'; got <sqlalchemy.ext.hybrid.hybrid_property object at ...>.

**Reply by zzzeek** [reply]
a foreign key refers to individual columns in the table, the reason is because foreign key indicates that a primary key value from the parent table will be going into that column.  so it's for persistence and can't be a composed expression.
